### PR TITLE
Add Github Action for building Docker images with R version extraction for tags

### PR DESCRIPTION
**Pull request by almahmoud**
I started this as a way to keep different R version containers for Bioconductor use since we need containers per R version, but I figured it might be a useful contribution upstream

This will rebuild the r2u Docker images every 2 days or any commit to a Dockerfile, and push with tags `[ubuntu name]` `[ubuntu name]-r-[rversion]` `[ubuntu version]` `[ubuntu version]-r-[rversion]` (see example in screenshot) 
image

without pinning the version in the script, by extracting the versions after the fact to push each image with tags based on both ubuntu and r version.
While arm64 builds are currently failing (I think due to R version mismatches), this design allows each image be built separately and pushed by digest first, then merged into tags allowing for various platforms to be pushed under the same tag once available.

I am also happy to add Dockerhub target to this so they get built together, if you tell me what secrets to expect credentials under.

Can also keep this as a Bioconductor internal thing if you're not interested.

### CODE REVIEW

**Comment by eddelbuettel** [comment]
It's late, it has been a long day, and I am lost so can we start at square one:  What is "GHCR" and why should r2u care?

r2u uses binaries from PPM.  And fills in where missing (ie BioC).  So if you want BioC on aarch64, say, then maybe we should tie this up from somewhere.  Might be easier to start on a call ...

**Comment by almahmoud** [comment]
Sorry about the timing, I am used to working across many timezones, and deal with things as they come up and as I have the energy, but please do not feel pressured to respond now if inconvenient for you, I am not expecting you to in any way!

And happy to set up a call and talk about tentative plans if that helps.

TLDR: In concrete terms, this could complement or replace the Dockerhub automated builds, to offer containers that are tagged `rocker/r2u:jammy` and `rocker/r2u:jammy-r-4.3.1`, so when a new R is available and the new `rocker/r2u:jammy` rebuilds, it starts being equivalent to `rocker/r2u:jammy-r-4.3.2`, but we keep the old tag `rocker/r2u:jammy-r-4.3.1`.

I mostly was interested in having containers with R versions that we could pin per Bioconductor version for our build process. You pointed us to the Dockerhub containers (which I assume are being built by Dockerhub's automated builds?), but right now if we point to `rocker/r2u:jammy`, it will have R 4.3.x today and R 4.4.x next year, which is not ideal for us since Bioc versions and R versions are not always in sync.

That being said, if you don't care for it on your side, you can also close the PR and I can leave it in a Bioconductor fork for us to use. I thought I'd PR upstream in case you found it useful, didn't mean to bother.

My main goal was to find the least intrusive way to keep old containers per R version, which this accomplishes in the github action build workflow without touching the Dockerfiles by extracting that information from the built container before tagging it. So it builds as it does now without pinning an R version, pushes the resulting container with it's digest hash without a tag, extracts the R version from within the built container, then tags it appropriately and pushes it there too.

Re the questions:
GHCR=GitHub Container Registry, an alternative to Dockerhub for hosting the containers as packages within the repository. Sorry for using the terminology without explaining. It is not central though, and could be entirely removed and replaced with just the Dockerhub registry as a target if you prefer. I always use GHCR as a first target instead of Dockerhub to push the containers because I can build and test all within github, and forks can inherit the build for their forks, so it's all easier for collaborative working with containers without needing any credentials.

We don't necessarily care about arm64 as a first pass, but we are already building binaries for arm64 for our containers which are available for both archs, so figured I'd keep that in mind as I make this quick docker build workflow so that leaves that door open with no extra work in the future when we want to start caring about that for bioc2u too. The failed arm64 builds are ignored anyway right now, I just thought I'd surface the error messages in the build, but I can remove if having the errored jobs with logs is not useful to anyone.

**Comment by eddelbuettel** [comment]
> GHCR=GitHub Container Registry

Ah yes, thank you. I once knew that too.  And upload one or two containers there when Docker started to be idiotic.

As for the rest maybe best to discuss about where it makes sense to stage things.  Maybe your project is mostly a "derivative of r2u that stands alone" for both arm64 and/or bioc.  I don;t have that much of a long-term plan for me this is a _small side project_ and I may only have time and energy for my core 'x86_64 binaries for cran with some bioc as needed'.  

And maybe the short term focus for Oct 1 should be to help you folks to produce some bioc binaries?

**Comment by almahmoud** [comment]
Yeh sounds like a quick zoom call would be useful.

I agree with the ideal goal of extending `r2u`, and I am hoping we can share the R layer for now, and ideally down the line maybe even have some shared dispatching/propagating code, so we have a joint effort keeping a shared stack up to date long term.

I don't think producing _some_ bioc binaries manually is an issue anymore, as we were able to manually dispatch to produce, and have collected system dependencies for all Bioc packages already too, so I think I am at the point where I am trying to dispatch the mechanism across all packages, and see how many work without special attention for potential corner cases.

I am ideally hoping to get to the point where I am giving a github repository the list of Bioc packages, and it builds all the `.deb` for each with free runners, self-dispatching itself in a cycle to constantly update, as an extension of what is currently happening for the container binaries (eg https://github.com/almahmoud/linux-amd64-317). This seemed like the best long-term solution that fits into something we're already maintaining for another project, and gives us frequent updates with no compute costs.

I was updating some of our containers anyway, so I figured I'd build these r2u containers with r versions "on the way" since I was in that headspace already and we had just talked about using these containers in Vince's repository, which long-term would need containers pinned by r version. bioc2u isn't my main focus yet (but will be soon) as I have some other work to wrap up first, but I had a burst of energy to move it a bit closer to where a first pass dispatch in Github Actions is possible, so these containers will be helpful for me even if we have them in a fork rather than merged upstream.

**Comment by eddelbuettel** [comment]
Yes, building via GH Actions is not a bad idea, and Inaki does that too for his rpm-based 'cran2copr' for Fedora.

I am really excited about what you and BioC do, and I think it is a wonderful and natural extension.  And a right next step.  I just want to be very careful about exactly what it is I "own" so that I do not bite off way more than I should and then add up with more than I can handle.  And adding another (full !!) binary build feels like too much for me right now.

But "we all", CRAN, ... do of course need arm64 binaries so I will help.  Chatting over zoom etc anytime may help. I'm in the Central time zone and should available most times.  I have a few meetings every now and then but general availability.

**Comment by eddelbuettel** [comment]
Yes, building via GH Actions is not a bad idea, and Inaki does that too for his rpm-based 'cran2copr' for Fedora.

I am really excited about what you and BioC do, and I think it is a wonderful and natural extension.  And a right next step.  I just want to be very careful about exactly what it is I "own" so that I do not bite off way more than I should and then add up with more than I can handle.  And adding another (full !!) binary build feels like too much for me right now.

But "we all", CRAN, ... do of course need arm64 binaries so I will help.  Chatting over zoom etc anytime may help. I'm in the Central time zone and should available most times.  I have a few meetings every now and then but general availability.

(Typed this hours ago and never hit the green 'comment' button. Sorry about that.)

**Comment by wybert** [comment]
This is something I am looking for. Any updates guys?

**Comment by almahmoud** [comment]
It's up to @eddelbuettel if this seems useful for this upstream repository. It might make sense to close it if there is not intention of merging.
We are using the contents of this PR at our Bioconductor fork: https://github.com/bioconductor/r2u , as we need the R versions, so in the meantime you may use containers from there with format `ghcr.io/bioconductor/r2u:jammy-r-4.3.1` if you want pinned R versions. They are rebuilding every 2 days in that repository, using the same script and Dockerfiles as this repository, and just adding the github action workflow for the GHCR push and R version tagging, as this PR does

**Comment by eddelbuettel** [comment]
I haev no immediate plans to expand this. I think I can commit to maintaining the two most recent LTS versions with the corresponding current R version.  I am doing this as a side project and I do not have the resources to do full grids of, say, two times LTS times three R versions each.  Sorry.

I have been running R myself with whatever the current version is for well over two and a half decades.  "Works for me."  So if someone is interested in using r2u with all it has to offer, it may mean 'recent R'.  There are other systems that offer support for older R versions but whenever I spot check they seem to offer only source not binary.

I hope this addresses and answers the question.  If not please clarify what the ask is here,  And of course thanks for trying r2u, and offering feedback.

**Comment by almahmoud** [comment]
Honestly, I don't really understand why you're considering the Github Action as extra work, as it doesn't add new LTS or R versions, all it does is push Docker images to Github not just Dockerhub, and extract R version from the built image to add it to the tag (you don't have to manually change anything, it just looks within the container after the build), so that when the image gets rebuilt next year, it's not completely overwriting the old version. It overwrites the generic `jammy` or `focal` tags that you have now, but there would be a Docker image under `jammy-r-4.3.1` still available for those who might have workflows relying on an old version. It doesn't change anything but the tag of the Docker image, so working with older versions becomes possible.
Unless you're deleting old versions from the repository, I don't think anything has to change for this PR to be useful. If you just don't want a GitHub action because maintaining the YAML file seems like too much work, then I understand that, but I don't think this implies any extra work beyond the YAML workflow with some bash in it. And if it stops working in the future because of something breaking, you'd just get back to having only DockerHub images like you currently do.

Without changing anything, this PR makes the Docker images exist in GitHub and with a more precise tagging scheme.

P.S. it makes 0 difference to me if this is in the upstream repository, since we have a fork that we're using for our purposes anyway, I just wanted to at least re-clarify so that you're making an informed decision, as so far I've just heard relatively vague concerns of maintaining more, but it's not clear to me what part is the issue. Ultimately it's your decision, we have what we need in the fork, I just wanted to PR upstream as is generally my open-source practice for what I see as public/open collaborative projects, and if the workflow needs maintaining later I am more than happy to contribute to its maintenance (I will have to anyway in the fork). If you want this project to remain always yours alone with no contributions that's also a valid way of doing things and I understand that as well, in which case closing this makes sense and we can just develop in parallel in the Bioc fork.

**Comment by eitsupi** [comment]
I have a simple question, but even if we use a container from the past, isn't it meaningless because when we try to install something (R package) on that container, it connects to the latest r2u repository?

**Comment by almahmoud** [comment]
I think it depends whether old versions are being deleted. If they are, that might change our `bioc2u` project as well, and you would be right. My assumption was that they aren't. In that case, you would have an old container with a specific version of R, and in theory packages that require a newer version won't install, so it should be pulling the latest version compatible with your version of R, which is why I think it would just work out of the box.
I very well might be missing something though.

**Comment by eddelbuettel** [comment]
Old versions have always been deleted and r2u makes no claim to be offering a time machine.  We can install _here_ and _now_ what is available today.

If you want to have the ability to install what we have at dat _x_ and a future day _x + y_ I fear you (or someone else) would be to add a caching layer, find a mirror, deal with the indices.  It call all be done in theory.  But it is not on my immediate or medium-term map. And I don't think I claimed I would offer this anytime soon.

What we have now is good.  I am all for setting a date for interested people (say @almahmoud @eitsupi @eddelbuettel) who have some experience to get together via Zoom to chat about a) what would be desirable b) what would be needed and c) what would be feasible next.  We are doing really well extending r2u 'horizontally' with bioc2u.  We may think about doing it 'vertically' across time too.  There may even be ways to use PPM for it, at least at snapshot dates.

**Comment by almahmoud** [comment]
My tentative approach with bioc2u is keeping all versions of packages, using https://github.com/deb-s3/deb-s3 to push at the end of a build cycle. So it will work there. I wrongly assumed r2u was adding new versions but keeping old ones too. Since r2u doesn't keep old packages and only works on latest version, it makes sense to only keep one container.
